fix: merge generation of input-data for `issue_closures` and `list_repos`

chazkiker2 opened the pull request:
### Description

- [x] fix #64 
- [x] combine the `Producer` functionality for `issue_closures` and `list_repos` into one producer
  - [x] fix associated consumers
- [x] add some clarifying comments & doc-comments throughout


r? @rylev

rylev commented on src/report/issue_closure.rs:
```diff
@@ -1,12 +1,14 @@
-use std::path::Path;
-
 #[derive(Debug, Serialize, Deserialize)]
 pub(super) struct IssueClosure {
```
We don't need this struct at all any more. We just need the `write_issue_closures` function which can read info from `ReportData.repo_infos`

rylev reviewed (changes requested):
This is definitely a good step in the right direction. I think we don't need there to always be a 1 to 1 relationship between input and output. We just need the repo infos input (and remove the idea of an issue closures input) which the issue closures output uses for the information it needs.

chazkiker2 commented on src/metrics/list_repos.rs:
```diff
@@ -48,10 +58,66 @@ impl Producer for ListReposForOrg {
+    start_date: &Datetime,
+    end_date: &Datetime,
+) -> IssueClosuresCount {
+    let opened = util::count_issues(
```
I'd be happy for these to run in parellel; 

is this as simple as `tokio::spawn` and `clone` any necessary data?

chazkiker2 replied:
gave this a shot... you'll have to let me know if I missed the mark :)

chazkiker2 commented on src/report/issue_closure.rs:
```diff
@@ -60,4 +61,21 @@ impl Report {
+    pub(super) async fn issue_closures(&self, _config: &ReportConfig) -> Vec<IssueClosure> {
+        debug!("Finding issue closures...");
+        // path to relevant input data
+        let repo_info = self.input_dir().join("repo-infos.csv");
```
agreed that this was not ideal naming. But now that we removed the `IssueClosure` struct, maybe the `repo-infos` name feels more appropriate?

Happy to rename it, too, if you have any ideas.

chazkiker2 commented:
Changes since last review: 

- [x] removed `IssueClosure` struct
- [x] fixed comment nits
- [x] attempted parallelization of the various count queries... but hoping to have some more experienced eyes on that :smile:



### `--replay-graphql` bug
- I found an issue with the `--replay-graphql`... not sure if this is a bug I introduced here or if it existed. 
- still need to dig deeper and figure out why this is happening
- the changes in this branch are working great without the `--replay-graphql` option; but as soon we run it WITH that flag, the GQL query is loading response data from the same file for both `metrics::util::count_pull_requests` and `metrics::util::count_issues`
 - running with `RUST_LOG=debug` demonstrates this issue well

    <details>
    
    <summary>expand to see logging output</summary>
        
    ```bash
    ❯ RUST_LOG=debug cargo run -- report data/optopodi/2021-07-31                 
       Compiling optopodi v0.1.0 (/Users/chazadmin/code/opensource/github-metrics)
        Finished dev [unoptimized + debuginfo] target(s) in 7.84s
         Running `target/debug/optopodi report data/optopodi/2021-07-31`
    [2021-07-31T18:13:32Z DEBUG reqwest::connect] starting new connection: https://api.github.com/
    [2021-07-31T18:13:32Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG optopodi::metrics::list_repos] Retried issue closure info for optopodi/optopodi: IssueClosuresCount { opened: 2, closed: 5 }
    ```
    
    ```shell
    
    ❯ RUST_LOG=debug cargo run -- --replay-graphql report data/optopodi/2021-07-31
        Finished dev [unoptimized + debuginfo] target(s) in 0.15s
         Running `target/debug/optopodi --replay-graphql report data/optopodi/2021-07-31`
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/all-repos/0.json` rather than github
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-participants/0.json` rather than github
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-infos/0.json` rather than github
    [2021-07-31T18:13:35Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-infos/0.json` rather than github
    [2021-07-31T18:13:35Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-infos/0.json` rather than github
    [2021-07-31T18:13:35Z DEBUG optopodi::metrics::list_repos] Retried issue closure info for optopodi/optopodi: IssueClosuresCount { opened: 5, closed: 5 }
    ```
        
    </details>

chazkiker2 commented:
is it possible that would happen due to the multi-threaded context?

rylev commented on src/metrics/list_repos.rs:
```diff
@@ -32,26 +34,119 @@ impl ListReposForOrg {
 }
 
+impl ListReposForOrg {
+    fn make_repo(&self, repo_name: &str) -> Repo {
```
I think `to_repo` would be a more appropriate name

rylev commented on src/metrics/list_repos.rs:
```diff
@@ -32,26 +34,119 @@ impl ListReposForOrg {
+    #[throws]
+    async fn spawn_count_pulls(&self) -> usize {
+        let mut repo = self.clone();
+        tokio::spawn(async move { repo.count_pulls().await }).await??
```
This doesn't really help: we spawn the task and then immediately await the result. This is analogous to spawing and thread and immediately joining on that thread. 

I think we should back this out and just get the pulls count and issue closures count one after the other.

rylev commented on src/metrics/list_repos.rs:
```diff
@@ -32,26 +34,119 @@ impl ListReposForOrg {
+        let opened = tokio::spawn(async move { repo.count_issues("created").await }).await??;
+
+        let mut repo = self.clone();
+        let closed = tokio::spawn(async move { repo.count_issues("closed").await }).await??;
```
As with the above, this still means that the opened and closed counts are retrieved in order. Let's removing the spawning for now.

rylev reviewed (changes requested):
It looks like the graphql caching does not take arguments into account so when we make the same request (a GitHub query) with a different query string, it reads the same response even though the request was different (because they query strings were different.

chazkiker2 commented on src/metrics/list_repos.rs:
```diff
@@ -32,26 +34,116 @@ impl ListReposForOrg {
+
+        let futures = vec![repo.count_issues("created"), clone.count_issues("closed")];
+
+        let resolved = futures::future::try_join_all(futures).await?;
+
+        let result = IssueClosuresCount {
+            opened: resolved[0],
+            closed: resolved[1],
+        };
```
@rylev [offered nit](https://github-metrics.zulipchat.com/#narrow/stream/291433-general/topic/issue.20closures.20--.20PR.20.2366/near/247940202):
>  Nit: you can just deconstruct opened and closed in the line above (without needing to first bind to resolved)

Though I had trouble achieving this due to the fact that `resolved` is of type `Vec<usize>` and the compiler refused to deconstruct into a tuple.

```rust
// resolved is type `Vec<usize>`
let resolved = futures::future::try_join_all(futures).await?;
// compiler expected struct `Vec<usize>`, found tuple `(_, _)`
let (opened, closed) = resolved;
```

Would love to include a fix to this nit, if anyone has suggestions.

rylev commented on src/metrics/list_repos.rs:
```diff
@@ -32,26 +35,115 @@ impl ListReposForOrg {
+        let mut repo = self.clone();
+        let mut clone = self.clone();
+
+        let futures = vec![repo.count_issues("created"), clone.count_issues("closed")];
```
I don't want to delay this any more, but wouldn't [`try_join`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.4/futures/future/trait.TryFutureExt.html#method.try_join) work?

chazkiker2 replied:
I do like that better 😄 . pushed up that change

rylev reviewed (approved):
Looks great! 🎉

chazkiker2 commented:
@bors r=rylev

rylev commented:
@bors r+
